pyro/infer/predictive.py

```diff
@@ -0,0 +1,170 @@
+    reshaped_samples = {}
+
+    for name, sample in posterior_samples.items():
+
+        batch_size, sample_shape = sample.shape[0], sample.shape[1:]
+
+        if num_samples is None:
+            num_samples = batch_size
+
+        elif num_samples != batch_size:
+            warnings.warn("Sample's leading dimension size {} is different from the "
+                          "provided {} num_samples argument. Defaulting to {}."
+                          .format(batch_size, num_samples, batch_size), UserWarning)
+            num_samples = batch_size
+
+        sample = sample.reshape((num_samples,) + (1,) * (max_plate_nesting - len(sample_shape)) + sample_shape)
+        reshaped_samples[name] = sample
+
+    if num_samples is None:
+        raise ValueError("No sample sites in model to infer `num_samples`.")
+
+    return_site_shapes = {}
+    for site in model_trace.stochastic_nodes + model_trace.observation_nodes:
+        site_shape = (num_samples,) + model_trace.nodes[site]['value'].shape
+        if isinstance(return_sites, (list, tuple, set)):
+            if site in return_sites:
+                return_site_shapes[site] = site_shape
+        else:
+            if (return_sites is not None) or (site not in reshaped_samples):
+                return_site_shapes[site] = site_shape
+
+    # handle _RETURN site
+    if isinstance(return_sites, (list, tuple, set)) and '_RETURN' in return_sites:
+        value = model_trace.nodes['_RETURN']['value']
+        shape = (num_samples,) + value.shape if torch.is_tensor(value) else None
+        return_site_shapes['_RETURN'] = shape
+
+    if not parallel:
+        return _predictive_sequential(model, posterior_samples, args, kwargs, num_samples,
+                                      return_site_shapes.keys(), return_trace)
+
+    def _vectorized_fn(fn):
+        """
+        Wraps a callable inside an outermost :class:`~pyro.plate` to parallelize
+        sampling from the posterior predictive.
+
+        :param fn: arbitrary callable containing Pyro primitives.
+        :return: wrapped callable.
+        """
+
+        def wrapped_fn(*args, **kwargs):
+            with pyro.plate("_num_predictive_samples", num_samples, dim=-max_plate_nesting-1):
+                return fn(*args, **kwargs)
+
+        return wrapped_fn
+
+    trace = poutine.trace(poutine.condition(_vectorized_fn(model), reshaped_samples))\
+        .get_trace(*args, **kwargs)
+
+    if return_trace:
+        return trace
+
+    predictions = {}
+    for site, shape in return_site_shapes.items():
+        value = trace.nodes[site]['value']
+        if site == '_RETURN' and shape is None:
+            predictions[site] = value
+            continue
+        if value.numel() < reduce((lambda x, y: x * y), shape):
+            predictions[site] = value.expand(shape)
+        else:
+            predictions[site] = value.reshape(shape)
+
+    return predictions
+
+
+def predictive(model, posterior_samples, *args, **kwargs):
```

Comment:
Since we are moving this, I think we can also clean up the interface as follows. WDYT?

```
def predictive(model, posterior_samples=None, guide=None, return_trace=False, return_sites=(), 
               parallel=False, model_args=(), model_kwargs={})
```

Then, we don't need to pass in `posterior_samples=None` for SVI and can just pass the guide. I think this is also one of those cases where its just cleaner to pass the `model_args` and `model_kwargs` separately rather than using varargs (code editors / notebooks will display all the arguments available).

Reply:
I have gone back and forth to find an easy-to-interpret args/kwargs for this utility, but none of them satisfied me. :( Regarding `model_args`, `model_kwargs`, I think we should get more opinions about it. The problem is most keywords are not necessary, but model `args/kwargs` are required to have. I think it is also good to make `predictive` a class with a method `get_samples` to return samples given particular args, kwargs; and another method `get_trace` to return traces. WDYT?

Reply:
> The problem is most keywords are not necessary, but model args/kwargs are required to have. 

kwargs won't be required - by default it will be `{}` or `None`. Same with args - by default it will be an empty tuple - `()`.

> I think it is also good to make predictive a class with a method get_samples to return samples given particular args, kwargs; and another method get_trace to return traces. WDYT?

I think this is a better option. Lets do this instead.

Reply:
Yeah, making it a class looks good. I don't feel it messy anymore. The tests look cleaner than before. :)